src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs

```diff
@@ -738,19 +742,35 @@ override public object ExecuteScalar()
             // between entry into Execute* API and the thread obtaining the stateObject.
             _pendingCancel = false;
 
-            SqlStatistics statistics = null;
-
+            Guid operationId = _diagnosticListener.WriteCommandBefore(this, nameof(ExecuteScalar));
 
+            SqlStatistics statistics = null;
+            
+            Exception e = null;
             try
             {
                 statistics = SqlStatistics.StartTimer(Statistics);
                 SqlDataReader ds;
                 ds = RunExecuteReader(0, RunBehavior.ReturnImmediately, returnStream: true);
                 return CompleteExecuteScalar(ds, false);
             }
+            catch (Exception ex)
+            {
+                e = ex;
+                throw;
+            }
             finally
             {
                 SqlStatistics.StopTimer(statistics);
+
+                if (e != null)
+                {
+                    _diagnosticListener.WriteCommandError(operationId, this, nameof(ExecuteScalar), e);
+                }
+                else
+                {
+                    _diagnosticListener.WriteCommandAfter(operationId, this, nameof(ExecuteScalar));
```

Comment:
@stephentoub  Thanks for the review :smile:  
I think it makes more sense to have two different methods in this case. 

@tstringer Can you use [CallerMemberName]  as suggested by @stephentoub . It keeps the code cleaner.